File: real-time-canvas-web/types/room.ts
/**
 * Room management types
 * Defines room data structures and authentication types
 */

/**
 * Room role for users
 */
export type RoomRole = 'owner' | 'editor' | 'viewer'

/**
 * Room user with role
 */
export interface RoomUser {
  userId: string
  username: string
  role: RoomRole
  joinedAt: Date
  lastActiveAt: Date
}

/**
 * Room data structure
 */
export interface Room {
  id: string
  name: string
  ownerId: string
  createdAt: Date
  updatedAt: Date
  users: RoomUser[]
  isPrivate: boolean
  inviteCode?: string
  settings: RoomSettings
  objectCount: number
  lastActivity: Date
}

/**
 * Room settings
 */
export interface RoomSettings {
  allowGuestUsers: boolean
  allowObjectEditing: boolean
  allowPhysics: boolean
  maxUsers: number
  canvasBackground: string
}

/**
 * Room creation request
 */
export interface CreateRoomRequest {
  name: string
  username: string
  isPrivate?: boolean
  settings?: Partial<RoomSettings>
}

/**
 * Room join request
 */
export interface JoinRoomRequest {
  roomId: string
  username: string
  inviteCode?: string
}

/**
 * Room invitation
 */
export interface RoomInvitation {
  id: string
  roomId: string
  code: string
  createdBy: string
  createdAt: Date
  expiresAt: Date
  usedBy?: string
  usedAt?: Date
  maxUses: number
  useCount: number
}

/**
 * Room list response
 */
export interface RoomListResponse {
  rooms: Room[]
  total: number
  page: number
  limit: number
}

/**
 * Guest user session
 */
export interface GuestSession {
  userId: string
  username: string
  roomId: string
  joinedAt: Date
  expiresAt: Date
}

/**
 * Authentication state
 */
export interface AuthState {
  isAuthenticated: boolean
  userId: string | null
  username: string | null
  roomId: string | null
  guestMode: boolean
  token?: string
}

File: real-time-canvas-service/internal/models/dto/room.go
package dto

import (
	"time"
)

// RoomResponse is the room data sent to clients
type RoomResponse struct {
	ID          string                 `json:"id"`
	Name        string                 `json:"name"`
	OwnerID     string                 `json:"ownerId"`
	IsPrivate   bool                   `json:"isPrivate"`
	InviteCode  string                 `json:"inviteCode,omitempty"`
	MaxUsers    int                    `json:"maxUsers"`
	ObjectCount int                    `json:"objectCount"`
	UserCount   int                    `json:"userCount"`
	LastActive  string                 `json:"lastActive"`
	CreatedAt   string                 `json:"createdAt"`
	UpdatedAt   string                 `json:"updatedAt"`
	Users       []UserResponse         `json:"users,omitempty"`
	Settings    map[string]interface{} `json:"settings,omitempty"`
}

// CreateRoomRequest represents room creation request
type CreateRoomRequest struct {
	Name      string `json:"name" binding:"required,min=3,max=100"`
	IsPrivate bool   `json:"isPrivate"`
	MaxUsers  int    `json:"maxUsers" binding:"omitempty,min=1,max=200"`
}

// JoinRoomRequest represents room join request
type JoinRoomRequest struct {
	RoomID     string `json:"roomId" binding:"required"`
	InviteCode string `json:"inviteCode"`
}

// UpdateRoomRequest represents room update request
type UpdateRoomRequest struct {
	Name     string `json:"name" binding:"omitempty,min=3,max=100"`
	MaxUsers int    `json:"maxUsers" binding:"omitempty,min=1,max=200"`
}

// ToRoomResponse converts Room to RoomResponse
func ToRoomResponse(
	id, name, ownerID string,
	isPrivate bool,
	inviteCode *string,
	maxUsers, objectCount, userCount int,
	lastActive, createdAt, updatedAt time.Time,
	users []UserResponse,
) RoomResponse {
	inviteCodeStr := ""
	if inviteCode != nil {
		inviteCodeStr = *inviteCode
	}

	return RoomResponse{
		ID:          id,
		Name:        name,
		OwnerID:     ownerID,
		IsPrivate:   isPrivate,
		InviteCode:  inviteCodeStr,
		MaxUsers:    maxUsers,
		ObjectCount: objectCount,
		UserCount:   userCount,
		LastActive:  lastActive.Format(time.RFC3339),
		CreatedAt:   createdAt.Format(time.RFC3339),
		UpdatedAt:   updatedAt.Format(time.RFC3339),
		Users:       users,
	}
}
